README.org
#+title: Readme

* What

This is the mono repository for my DevOps application.


gh-webhook/CHANGELOG.md
# Changelog

`gh-webhook` uses [PVP Versioning][1].
The changelog is available [on GitHub][2].

## 0.0.0.0

* Initially created.

[1]: https://pvp.haskell.org
[2]: https://github.com/hughjfchen/gh-webhook/releases


gh-webhook/test/Spec.hs
{-# LANGUAGE OverloadedStrings #-}
module TestMain (main) where

import Test.Syd

import Core.Types
import Has

-- | Following is only for example
-- | you must adapt it accordingly
main :: IO ()
main = sydTest $
  describe "test gh-webhook properties" $ do
    it "test property - property 1" $ True


frontend/src/jobs.js
// in jobs.js
import * as React from "react";
import {
    List,
    Datagrid,
    TextField,
    DateField,
    FileField,
    Create,
    SimpleForm,
    RadioButtonGroupInput,
  FileInput,
  required,
  choices,
} from 'react-admin';
import UrlFieldWithCustomLinkText from './UrlFieldWithCustomLinkText';
import UrlFieldWithLastFileNameAsLinkText from './UrlFieldWithLastFileNameAsLinkText';

         // <UrlFieldWithLastFileNameAsLinkText source="payload.contents.contents" download={true} label="Dump File"/>
         // <UrlFieldWithCustomLinkText source="last_update.report_url" linkText="Report" target="_blank" label="Report"/>
         // <FunctionField render={record => `${record.payload.contents.contents.split('/').pop()}`} label="Dump File"/>
         // <TextField source="payload.contents.contents" label="Dump File"/>
         // <DumpFileField source="payload.contents.contents" label="Dump File"/>
export const JobList = () => (
    <List>
       <Datagrid>
         <TextField source="payload.tag" label="Parse Type"/>
         <UrlFieldWithLastFileNameAsLinkText source="payload.contents.contents" download={true} label="Dump File"/>
         <DateField source="created_at" showTime={true} locales='zh-CN' options={{hour12: false}}/>
         <DateField source="run_at" showTime={true} locales='zh-CN' options={{hour12: false}}/>
         <DateField source="updated_at" showTime={true} locales='zh-CN' options={{hour12: false}}/>
         <TextField source="status" />
         <UrlFieldWithCustomLinkText source="last_update.report_url" linkText="Report" target="_blank" label="Report"/>
       </Datagrid>
   </List>
);

const validateParseType = [required(), choices(['ParseJavaCore', 'ParseHeapDump'], 'Please choose one of the values')];
const validateDumpFile = [required()]
const tagJobCreateReq = data => ({
  ...data,
  createReq: "JobCreateReq"
});

export const JobCreate = props => (
     <Create {...props} transform={tagJobCreateReq} >
        <SimpleForm>
          <RadioButtonGroupInput source="parsetype" choices={[
            { id: 'ParseJavaCore', name: 'Java Core' },
            { id: 'ParseHeapDump', name: 'Heap Dump' },
          ]} validate={validateParseType}/>
          <FileInput source="file" label="Java Dump File" multiple={false} minSize={0} maxSize={1000000000} placeholder={<p>Drop your java dump file here</p>} validate={validateDumpFile}>
            <FileField source="src" title="title" />
          </FileInput>
        </SimpleForm>
     </Create>
);


gh-webhook/app/Main.hs
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE OverloadedStrings #-}

-- | This is the entry point to the main program
module Main (main) where

import AppCapability.DumpFetchor ()
import AppCapability.JavaAnalyzerRunner ()
import AppCapability.ReportPostProcessor ()
import AppCapability.ReportPreProcessor ()
import AppCapability.ReportUploader ()
import AppEnv
import AppM
-- import OddJobs.Migrations (createJobTable)

-- Note: It is not necessary to use fast-logger. You can use any logging library
-- that can give you a logging function in the IO monad.

import qualified Conferer as Conf
import Core.JavaAnalyzerRunner
import Core.Types
import qualified Data.ByteString as BS
import Data.Pool (Pool, createPool, destroyAllResources)
import qualified Database.PostgreSQL.Simple as PGS
import FromConfig.ResourcePool (PoolInfo (..))
import OddJobs.Cli (CliType (..), CommonStartArgs (startPidFile), defaultStartCommand, defaultStopCommand)
import OddJobs.ConfigBuilder (defaultJobType, defaultLogStr, defaultTimedLogger, mkConfig)
import OddJobs.Job (Config (..), Job (..), startJobRunner, throwParsePayload)
import qualified OddJobs.Types as OT
import System.FilePath (takeDirectory, (</>))
import System.Log.FastLogger (LogType' (..), defaultBufSize, withTimedFastLogger)
import System.Log.FastLogger.Date (newTimeCache, simpleTimeFormat)
import UnliftIO (MonadUnliftIO, bracket, withRunInIO)

-- | Copied from the OddJobs to config the pool. Convenience function to create a DB connection-pool with some sensible
-- defaults. Please see the source-code of this function to understand what it's
-- doing.
withConnectionPoolWithConfig ::
  (MonadUnliftIO m) =>
  Either BS.ByteString PGS.ConnectInfo ->
  PoolInfo ->
  (Pool PGS.Connection -> m a) ->
  m a
withConnectionPoolWithConfig connConfig poolInfo action = withRunInIO $ \runInIO ->
  bracket poolCreator destroyAllResources (runInIO . action)
  where
    poolCreator = liftIO $
      case connConfig of
        Left connString ->
          createPool (PGS.connectPostgreSQL connString) PGS.close (poolStripe poolInfo) (poolIdleTime poolInfo) (poolSize poolInfo)
        Right connInfo ->
          createPool (PGS.connect connInfo) PGS.close (poolStripe poolInfo) (poolIdleTime poolInfo) (poolSize poolInfo)

myJobRunner :: AppEnv' -> Job -> IO ()
myJobRunner env' job =
  throwParsePayload job >>= \case
    ParseJavaCore javaCorePath ->
      runAppM' env' $ genJavaCoreReport (jobId job) javaCorePath
    ParseHeapDump heapDumpPath ->
      runAppM' env' $ genHeapDumpReport (jobId job) heapDumpPath
    ParseGC gcLogPath ->
      runAppM' env' $ genGCReport (jobId job) gcLogPath

setOJDefaultTimeout :: AppEnv -> OT.Config -> OT.Config
setOJDefaultTimeout appEnv inConfig =
  inConfig
    { cfgDefaultJobTimeout = cfgDefaultJobTimeout $ appEnvOddJobsConfig appEnv,
      cfgPollingInterval = cfgPollingInterval $ appEnvOddJobsConfig appEnv,
      cfgDefaultMaxAttempts = cfgDefaultMaxAttempts $ appEnvOddJobsConfig appEnv
    }

main :: IO ()
main = do
  config <- Conf.mkConfig "JavaAnalyzerRunner"
  appEnv <- Conf.fetch config :: IO AppEnv

  -- check config data
  -- print $ appEnvDatabase appEnv
  -- print $ appEnvPool appEnv
  -- print $ appEnvOddJobsStartArgs appEnv
  -- print $ appEnvOddJobsStopArgs appEnv
  -- print $ appEnvOddJobsConfig appEnv
  -- print $ cfgTableName $ appEnvOddJobsConfig appEnv
  appEnv' <- runAppM appEnv configIn

  case appEnvCommand appEnv of
    Start -> defaultStartCommand (appEnvOddJobsStartArgs appEnv) Nothing $ CliOnlyJobRunner {cliStartJobRunner = startJobMonitor appEnv appEnv'}
    Stop -> defaultStopCommand (appEnvOddJobsStopArgs appEnv)
    Status -> error "The Status command is not implemented yet."
    Unknown -> error "Unknown command, it must be Start, Stop or Status."
  where
    -- A callback-within-callback function. If the commands-line args contain a
    -- `start` command, this function will be called. Once this function has
    -- constructed the 'Config' (which requires setting up a logging function,
    -- and a DB pool) it needs to execute the `callback` function that is passed
    -- to it.
    startJobMonitor theEnv theEnv' overrideFn =
      -- a utility function provided by `OddJobs.ConfigBuilder` which ensures
      -- that the DB pool is gracefully destroyed upon shutdown.
      withConnectionPoolWithConfig (Right $ appEnvDatabase theEnv) (appEnvPool theEnv) $ \dbPool -> do
        -- setup db
        -- withResource dbPool $ \conn -> createJobTable conn $ cfgTableName $ appEnvOddJobsConfig theEnv
        -- Boilerplate code to setup a TimedFastLogger (from the fast-logger library)
        tcache <- newTimeCache simpleTimeFormat
        withTimedFastLogger tcache (LogFileNoRotate (takeDirectory (startPidFile $ appEnvOddJobsStartArgs theEnv) </> "my-job-runner.log") defaultBufSize) $ \logger -> do
          -- Using the default string-based logging provided by
          -- `OddJobs.ConfigBuilder`. If you want to actually use
          -- structured-logging you'll need to define your own logging function.
          let jobLogger = defaultTimedLogger logger (defaultLogStr defaultJobType)
          startJobRunner $ mkConfig jobLogger (cfgTableName $ appEnvOddJobsConfig theEnv) dbPool (cfgConcurrencyControl $ appEnvOddJobsConfig theEnv) (myJobRunner theEnv') (overrideFn . setOJDefaultTimeout theEnv)


gh-webhook/cd/fix-up/do.sh
#!/usr/bin/env bash

if ! type dirname > /dev/null 2>&1; then
    echo "Not even a linux or macOS, Windoze? We don't support it. Abort."
    exit 1
fi

. "$(dirname "$0")"/../common/common.sh

init_with_root_or_sudo "$0"

SCRIPT_ABS_PATH=$(turn_to_absolute_path "$0")

begin_banner "Top level" "project deploy - fix-up"

# some command fix up for systemd service, especially web server
getent group nogroup > /dev/null || sudo groupadd nogroup

if [ -n "$NIX_STORE_PATH" ]; then
  if [ -n "$RELEASE_HAS_SYSTEMD_SERVICE" ] && [ "true" == "$RELEASE_HAS_SYSTEMD_SERVICE" ]; then
    sudo "$NIX_STORE_PATH"/bin/setup-systemd-units
  else
    info "To use the program, type $NIX_STORE_PATH/bin/gh-webhook-bin-sh at the command prompt."
  fi
else
  my_exit "cannot determine the build artifect path without the NIX_STORE_PATH environment variable."
fi

done_banner "Top level" "project deploy - fix-up"


api-gw/openresty/lua/parsedump.lua
local upload = require('resty.upload')
local cjson = require('cjson')

local chunk_size = 4096
local form ,err = upload:new(chunk_size)
if not form then
    ngx.log(ngx.ERR, "failed to new upload: ", err)
    ngx.exit(ngx.HTTP_INTERNAL_SERVER_ERROR)
end

form:set_timeout(5 * 60 * 1000)
local conf = {max_size=1000 * 1000000, allow_exts={'txt', 'threaddump', 'log', 'phd', 'hprof', 'zip', 'jar'}}

local my_my_upload_home = ngx.var.my_upload_home
local function get_upload_home(dRoot)
	if dRoot then
		return dRoot..'/dumpfiles'
	else
		return '/srv/dumpfiles'
	end
end
local upload_home = get_upload_home(my_my_upload_home)

local file
local file_name

local current_k
local current_v
local kv_part = {}

local function get_ext(res)
    local filename = ngx.re.match(res,'(.+)filename="(.+)\\.(.+)"(.*)')
    return filename[3]
end

local function check_upload_type(res)
    local ext
    if res == 'text/plain' then
        ext = 'txt'
    elseif res == 'application/java-archive' then
        ext = 'jar'
    elseif res == 'application/zip' then
        ext = 'zip'
    elseif res == 'application/octet-stream' then
        ext = 'phd'
    end
    return ext
end

local function in_array(v, tab)
    local i = false
    for _, val in ipairs(tab) do
        if val == v then
            i = true
            break
        end
    end
    return i
end

local function get_filename(res)
    local filename = ngx.re.match(res,'(.+)filename="(.+)"(.*)')
    if filename then
        return filename[2]
    end
end

local function get_kv_name(res)
    local kvname = ngx.re.match(res,'(.+)name="(.+)"(.*)')
    if kvname then
        return kvname[2]
    end
end

while true do
    local typ, res, err = form:read()
    if not typ then
        ngx.say(cjson.encode({code=502, msg='failed to read:', data=err}))
        return
    end
    if typ == "header" then
        if res[1] == "Content-Disposition" then
            local upload_file_name = get_filename(res[2])

	    if upload_file_name then  -- multipart the file part
		local extension = get_ext(res[2])
            	if not extension then
                 	ngx.say(cjson.encode({code=501, msg='ext not found, it must be txt, threaddump, log, jar, zip, hprof or phd.', data=res}))
                 	return
            	end

            	if not in_array(extension, conf.allow_exts) then
                	ngx.say(cjson.encode({code=501, msg='upload file type not supported, it must be txt, threaddump, log, jar, zip, hprof or phd.', data=res}))
                	return
            	end

            	local dir = upload_home..'/'..os.date('%Y%m%d')..'/'
            	local status = os.execute('mkdir -p '..dir)
            
		if not status then
                	ngx.say(cjson.encode({code=501, msg='mkdir for the upload home failed'}))
                	return
            	end
            	file_name = dir..upload_file_name
            	if file_name then
                	file = io.open(file_name, "w+")
                	if not file then
                    		ngx.say(cjson.encode({code=500, msg='failed to create file for the upload file',data=res}))
                    		return
                	end
            	end
	    else  -- multipart non-file part
		current_k = get_kv_name(res[2])
	    end
	elseif res[1] == "Content-Type" then
            local upload_file_type = check_upload_type(res[2])
            if not upload_file_type then
                 ngx.say(cjson.encode({code=501, msg='cloud not determine the upload file type. It must be txt, threaddump, log, jar, zip, hprof or phd.', data=res}))
                 return
            end
            if not in_array(upload_file_type, conf.allow_exts) then
           	ngx.say(cjson.encode({code=501, msg='upload file type not supported with current content-type, it must be txt, threadump, log, jar, zip, hprof or phd.', data=res}))
           	return
            end
        end
     elseif typ == "body" then
        if type(tonumber(res)) == 'number' and tonumber(res) > conf.max_size then
            ngx.say(cjson.encode({code=501, msg='oversize', data=res}))
            return
        end
        if file then
            file:write(res)
	else
	    current_v = res
	    kv_part[current_k] = current_v
        end
    elseif typ == "part_end" then
        if file then
            file:close()
            file = nil
        end
    elseif typ == "eof" then
	-- upload finished, now call postgrest to insert a record into db to kickstart a job to analyze the upload file.
	if not file_name then
	    ngx.status = ngx.HTTP_BAD_REQUEST
	    ngx.say("No file uploaded, if you're using curl command, you should make sure put a @ symbol before the file name.")
        else
            local file_uri = string.gsub(file_name, my_my_upload_home, '')
	    local download_url = ngx.var.scheme.."://"..ngx.var.http_host..file_uri
	    local job_create_payload = {tag=kv_part["parsetype"], contents={tag='HttpUrl', contents=download_url}}
	    local job_create_req = {status='queued', payload=job_create_payload}
	    ngx.log(ngx.INFO, cjson.encode(job_create_req))
	    ngx.req.set_header("Content-Type", "application/json")
	    ngx.req.set_header("Accept", "application/json")
	    ngx.req.set_header("Prefer", "return=headers-only")
            local job_create_res = ngx.location.capture("/rest/rpc/submit_parsedump_job", {method = ngx.HTTP_POST, body = cjson.encode(job_create_req)})
	    if job_create_res then
                ngx.status = job_create_res.status
	        ngx.header["Content-Type"] = job_create_res.header["Content-Type"]
                -- local id_location = ngx.re.match(job_create_res.header["Location"], '(.+)=eq\\.(.+)')
	        -- local theId
	        -- if id_location then
                --    theId = id_location[2]
	        -- end
	        -- ngx.header = job_create_res.header
                -- ngx.say(cjson.encode({ id = theId }))
                ngx.say(job_create_res.body)
	    end
	end
        break
    else

    end
end


api-gw/openresty/lua/uploadreport.lua
local upload = require('resty.upload')
local cjson = require('cjson')

local chunk_size = 4096
local form ,err = upload:new(chunk_size)
if not form then
    ngx.log(ngx.ERR, "failed to new upload: ", err)
    ngx.exit(ngx.HTTP_INTERNAL_SERVER_ERROR)
end

form:set_timeout(5 * 60 * 1000)
local conf = {max_size=1000 * 1000000, allow_exts={'htm', 'html', 'tgz'}}

local my_my_upload_home = ngx.var.my_upload_home
local function get_upload_home(dRoot)
	if dRoot then
		return dRoot..'/parsereports'
	else
		return '/srv/parsereports'
	end
end
local upload_home = get_upload_home(my_my_upload_home)

local file
local file_name

local current_k
local current_v
local kv_part = {}

local function get_ext(res)
    local filename = ngx.re.match(res,'(.+)filename="(.+)\\.(.+)"(.*)')
    return filename[3]
end

local function check_upload_type(res)
    local ext
    if res == 'text/html' then
        ext = 'html'
    elseif res == 'application/zip' then
        ext = 'zip'
    elseif res == 'application/octet-stream' then
        ext = 'tgz'
    end
    return ext
end

local function in_array(v, tab)
    local i = false
    for _, val in ipairs(tab) do
        if val == v then
            i = true
            break
        end
    end
    return i
end

local function get_filename(res)
    local filename = ngx.re.match(res,'(.+)filename="(.+)"(.*)')
    if filename then
        return filename[2]
    end
end

local function get_kv_name(res)
    local kvname = ngx.re.match(res,'(.+)name="(.+)"(.*)')
    if kvname then
        return kvname[2]
    end
end

while true do
    local typ, res, err = form:read()
    if not typ then
        ngx.say(cjson.encode({code=502, msg='failed to read:', data=err}))
        return
    end
    if typ == "header" then
        if res[1] == "Content-Disposition" then
            local upload_file_name = get_filename(res[2])

	    if upload_file_name then  -- multipart the file part
		local extension = get_ext(res[2])
            	if not extension then
                 	ngx.say(cjson.encode({code=501, msg='ext not found, it must be txt, jar, zip or phd.', data=res}))
                 	return
            	end

            	if not in_array(extension, conf.allow_exts) then
                	ngx.say(cjson.encode({code=501, msg='upload file type not supported, it must be htm, html or zip.', data=res}))
                	return
            	end

            	local dir = upload_home..'/'..os.date('%Y%m%d')..'/'
            	local status = os.execute('mkdir -p '..dir)
            
		if not status then
                	ngx.say(cjson.encode({code=501, msg='mkdir for the upload home failed'}))
                	return
            	end
            	file_name = dir..upload_file_name
            	if file_name then
                	file = io.open(file_name, "w+")
                	if not file then
                    		ngx.say(cjson.encode({code=500, msg='failed to create file for the upload file',data=res}))
                    		return
                	end
            	end
	    else  -- multipart non-file part
		current_k = get_kv_name(res[2])
	    end
	elseif res[1] == "Content-Type" then
            local upload_file_type = check_upload_type(res[2])
            if not upload_file_type then
                 ngx.say(cjson.encode({code=501, msg='cloud not determine the upload file type. It must be htm, html or zip.', data=res}))
                 return
            end
            if not in_array(upload_file_type, conf.allow_exts) then
           	ngx.say(cjson.encode({code=501, msg='upload file type not supported with current content-type, it must be htm, html or zip.', data=res}))
           	return
            end
        end
     elseif typ == "body" then
        if type(tonumber(res)) == 'number' and tonumber(res) > conf.max_size then
            ngx.say(cjson.encode({code=501, msg='oversize', data=res}))
            return
        end
        if file then
            file:write(res)
	else
	    current_v = res
	    kv_part[current_k] = current_v
        end
    elseif typ == "part_end" then
        if file then
            file:close()
            file = nil
        end
    elseif typ == "eof" then
	-- upload finished, now call postgrest to insert a record into db to save the url for the uploaded report.
	local report_path = file_name
	-- if the upload report is zip, we need to unzip it
	local file_ext = string.sub(file_name, -4, -1)
	if file_ext == ".tgz" then
	   local file_without_ext = string.sub(file_name, 1, string.len(file_name)-4)
	   local mk_report_dir_status = os.execute('mkdir -p '..file_without_ext)
	   if not mk_report_dir_status then
              ngx.say(cjson.encode({code=501, msg='mkdir for the upload zip report failed'}))
              return
           end
	   local unzip_report_status = os.execute('cd '..file_without_ext..' ; tar zxf '..file_name)
	   if not unzip_report_status then
              ngx.say(cjson.encode({code=501, msg='untar the upload tgz report failed'}))
              return
           end    
	   report_path=file_without_ext..'/index.html'
	end
        local file_uri = string.gsub(report_path, my_my_upload_home, '')
	local download_url = ngx.var.scheme.."://"..ngx.var.http_host..file_uri
	local report_create_req = { job_id = kv_part["job_id"]
	                          -- , created_at = kv_part["created_at"]
				  -- , updated_at = kv_part["updated_at"] -- use now() instead the old update time.
				  -- , run_at = kv_part["run_at"]
				  -- , status = kv_part["status"]
				  -- , payload = cjson.decode(kv_part["payload"])
				  , last_update = { report_url = download_url }
				  -- , attempts = kv_part["attempts"]
				  -- , locked_at = kv_part["locked_at"]
				  -- , locked_by = kv_part["locked_by"]
				  }
	ngx.log(ngx.INFO, cjson.encode(report_create_req))
	ngx.req.set_header("Content-Type", "application/json")
	ngx.req.set_header("Accept", "application/json")
	-- ngx.req.set_header("Prefer", "return=headers-only")
        local report_create_res = ngx.location.capture("/rest/rpc/update_parse_report", {method = ngx.HTTP_POST, body = cjson.encode(report_create_req)})
	if report_create_res then
            ngx.status = report_create_res.status
	    ngx.header["Content-Type"] = report_create_res.header["Content-Type"]
            -- local id_location = ngx.re.match(report_create_res.header["Location"], '(.+)=eq\\.(.+)')
	    -- local theId
	    -- if id_location then
            --    theId = id_location[2]
	    -- end
	    -- ngx.header = job_create_res.header
            -- ngx.say(cjson.encode({ id = theId }))
            ngx.say(report_create_res.body)
	end
        break
    else

    end
end
